src/mmocore/MMOClient.ts

```diff
@@ -10,7 +10,7 @@ import IMMOClientMutator from "./IMMOClientMutator";
 import AbstractPacket from "./AbstractPacket";
 import MMOConfig from "./MMOConfig";
 
-export default abstract class MMOClient implements IProcessable {
+export default abstract class MMOClient extends EventEmitter implements IProcessable {
```

Comment:
Good idea, but in that case I would put the `EventTypes.ts` folder within `mmocore`

Reply:
@npetrovski thank you for your feedback!

> The problem with this approach is the fact you gonna have 2 instances of EventEmitter - one for LoginClient, and another for the GameClient, aka. bigger memory footprint.

You are correct, but in practice, the LoginClient is shortlived, usually should live for couple seconds. We could make sure that after client is disconnected it's event handlers are detached.

> What if we keep the EventEmitter singleton (preserving export const GlobalEvents = new EventEmitter();) and simply add a new property at the abstract MMOClient?

Singleton does not help here, as same singleton is shared trough out multiple clients, which does not solve the issue of not sharing same event emitter trough multiple clients. **Key** here would be to get rid of all calls that use `GlobalEvents`.

If we would like to use single instance of _GlobalEvents_, we must then prefix every _EventHandler_ with a client instance.
eg.:
```ts
export type EventHandler = ((client: MMOClient, evt: Event) => void) & { _once?: boolean };
```

and then in every event handler user would have to talk to it's client explicitly.
eg.:
```ts
l2.on("LoggedIn", (client) => {
  const x = 50 + Math.floor(Math.random() * 50) + client.Me.X;
  const y = 50 + Math.floor(Math.random() * 50) + client.Me.Y;
  const z = client.Me.Z;
  client.moveTo(x, y, z);
});
```

Only this way, you are able to assign EventHandler to specific clients using GlobalEvents. I'd be worried that this approach would start to pollute the GlobalEvents handlers, as there would be only single instance for let's say realm of a 9 players (full party).

Reply:
I see your point, and I would rather agree with your 1st approach (`MMOClient extends EventEmitter`). Especially when it comes to this issue https://github.com/npetrovski/l2js-client/issues/27.

My only concern (apart from the memory footprint) is what do we do with [Client.ts](https://github.com/npetrovski/l2js-client/blob/master/src/Client.ts) event handler methods, or in other words - how do we fire or bind to a specific event from user point of view when a `new Client` is instantiated ? Should we change those methods as well (https://github.com/npetrovski/l2js-client/blob/554de4bd2c110edb672d8b51120ea4d86fbc0066/src/Client.ts#L69, https://github.com/npetrovski/l2js-client/blob/554de4bd2c110edb672d8b51120ea4d86fbc0066/src/Client.ts#L75, https://github.com/npetrovski/l2js-client/blob/554de4bd2c110edb672d8b51120ea4d86fbc0066/src/Client.ts#L81)?

Reply:
I have changed them already here:
https://github.com/npetrovski/l2js-client/pull/41/files#diff-b7961e132a9ada7fe5b4cdec69198fc7e4fe506e45b5a7fe29f399a01c45ff53L69-L85

I need to find smart way about login client events as now only game client events are handled, but unless you have used login client events outside of internal API everything should work as before (exceptions being `TODO`s).

There is also one more issue and that I am thinking of making either base L2Object or L2Creature class event emitter, so they can trigger events and then client could listen to them and forward them. As these classes do not have reference to the client we can't trigger events from them that client would listen for, adding this reference does not make much sense to me, but we could have owner property on everything that client holds and would solve the issue quite nicely.

Reply:
Yeah, my bad - I missed the `Client.ts` changes. Indeed, that was bothering me with that approach - the user should be able to also hook on the LoginClient events as well. It's all about the usage. In fact what I had in mind when added the `GlobalEvents` (as well as the Commands proxy) was to make the `LoginClient` and `GameClient` kinda transparent for the user - the user should be able to do everything from the `Client` exposed API directly.

Let me throw another idea. If I follow what you suggested (with the 1st approach) here, and in order to avoid having the `Client.ts` methods like `'onLoginClient()` and `onGameClient()` (or prefixes with all the events) and keep the `on(), once(), off()` method only, maybe we can introduce another property in the `Client.ts` called "ActiveClient", and switch the value of this property based on the login phase of the game. A.k.a first we can have the ActiveClient linked to LoginClient, and whenever the state of the app is changed, and the connection to LS is closed, we can switch the ActiveClient to the GameClient. What do you think?

Regarding the L2Object being also an event emitter - I agree. This will for sure make the project API much convenient to use.

Reply:
Yes you do follow it correctly.

I don't think `ActiveClient` property would work in this case, as even handlers are registered before you even initiate connection and that would register everything to most likely login client or nowhere as there wouldn't be active client set, so the current API would be broken.

I am thinking about registering the events on both clients

```ts
  on(...params: EventHandlerType): this {
    const c = this.___event_params(...params);
    const c = this.___event_params(...params);
    this.GameClient.on(c.type, c.handler);
    this.LoginClient && this.LoginClient.on(c.type, c.handler);
    return this;
  }
```

memory-wise we should be okay, as the event handler is shared between the clients, we will pay small price of another reference in the array and that could only apply to small period of time and events that are registered immediately.

To avoid the *small* overhead user could register the handlers once logged in:
```ts
l2.on('Logged', () => {
  l2.on('Attacked', () => {})
})
```
but personally I have feeling that it wouldn't be that much of difference in terms of memory.

Additional check could ensure that once logged in, we would not register events on the LoginClient anymore, and once LoginClient does its job, we can remove all the handlers from it and set LoginClient to null.

Reply:
I agree. Registering the events for both clients is a nice workaround, and we can start from that point. Could you please update the pull requests whenever you have the time, so that I can merge (I am also preparing some more fixes to push after).

br. It's a pleasure working with you!